src/server/supplements/recorderSupplement.ts

```diff
@@ -294,4 +315,50 @@ export class RecorderSupplement {
     const pageAlias = this._pageAliases.get(page)!;
     this._generator.signal(pageAlias, page.mainFrame(), { name: 'dialog', dialogAlias: String(++this._lastDialogOrdinal) });
   }
+
+  async onBeforeCall(sdkObject: SdkObject, metadata: CallMetadata): Promise<void> {
+    this._sdkObject = sdkObject;
+    this._callMetadata = metadata;
+    const { source, line } = this._source(metadata);
+    this._recorderApp?.setSource(source, 'javascript', line);
+    if (metadata.method === 'pause' || (this._pauseOnNextStatement && metadata.method === 'goto'))
+      await this.pause();
+  }
+
+  async onAfterCall(sdkObject: SdkObject, metadata: CallMetadata): Promise<void> {
+    this._sdkObject = null;
+    this._callMetadata = null;
+  }
+
+  async onBeforeInputAction(sdkObject: SdkObject, metadata: CallMetadata): Promise<void> {
+    if (this._pauseOnNextStatement)
+      await this.pause();
+  }
+
+  private _source(metadata: CallMetadata): { source: string, line: number | undefined } {
+    let source = '// No source available';
+    let line: number | undefined = undefined;
+    if (metadata.stack && metadata.stack.length) {
+      try {
+        source = this._readAndCacheSource(metadata.stack[0].file);
+        line = metadata.stack[0].line ? metadata.stack[0].line - 1 : undefined;
+      } catch (e) {
+        source = metadata.stack.join('\n');
```

Comment:
Stack is an array of structs, not strings, isn't it?

Reply:
Will follow up!